lib/auth/authentication_backend.rb

```diff
@@ -29,7 +24,130 @@ def authentication_flow(params)
+    raise AuthenticationBackendException.new("No Response Back from Authentication Service to process")
+  end
+
+  def create_group(name:, description:)
+    client.create_group(
+      group_name: name,
+      description: description,
+      user_pool_id: user_pool_id
+    )
+  rescue Aws::CognitoIdentityProvider::Errors::InvalidParameterException => e
+    raise AuthenticationBackendException.new(e.message)
+  rescue Aws::CognitoIdentityProvider::Errors::GroupExistsException => e
+    raise GroupExistsException.new(e.message)
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new(e.message)
+  end
+
+  # Returns a secret shared code to associate a TOTP app/device with
+  def associate_device(access_token:)
+    associate = client.associate_software_token(access_token: access_token)
+    associate.secret_code
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new("Error occurred associating device with error #{e.message}")
+  end
+
+  def enrole_totp_device(access_token:, totp_code:)
+    client.verify_software_token(
+      access_token: access_token,
+      user_code: totp_code
+    )
+    client.set_user_mfa_preference(
+      access_token: access_token,
+      software_token_mfa_settings: {
+        enabled: true,
+        preferred_mfa: true
+      }
+    )
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new(e.message)
+  end
+
+  def add_user(email:, given_name:, family_name:, roles:, phone_number: nil)
+    temporary_password = ""
+    until password_meets_criteria?(temporary_password) do
+      temporary_password = generate_password
+    end
+    client.admin_create_user(
+      temporary_password: temporary_password,
+      user_attributes: [
+        {
+          name: 'email',
+          value: email
+        },
+        {
+          name: 'given_name',
+          value: given_name
+        },
+        {
+          name: 'family_name',
+          value: family_name
+        },
+        {
+          name: 'phone_number',
+          value: phone_number
+        },
+        {
+          name: 'custom:roles',
+          value: roles.join(",")
+        }
+      ],
+      username: email,
+      user_pool_id: user_pool_id
+  )
+  rescue Aws::CognitoIdentityProvider::Errors::AliasExistsException,
+         Aws::CognitoIdentityProvider::Errors::UsernameExistsException => e
+    raise UsernameExistsException.new(e.message)
+  rescue StandardError => e
+    raise AuthenticationBackendException.new(e.message)
+  end
+
+  def add_user_to_group(username:, group:)
+    client.admin_add_user_to_group(
+      user_pool_id: user_pool_id,
+      username: username,
+      group_name: group
+    )
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new(e.message)
+  end
+
+  def get_users(limit: 50)
```

Comment:
I believe 60 is the maximum, we should use that, so we don't have to paginate for some time...
```suggestion
  def get_users(limit: 60)
```